rocBLAS: Add an option to disable tensile

teonnik opened the pull request:
- Tensile takes a lot of time to compile, add an option to disable it
- Change variant name to be consistent with rocsolver
- Change default variant value to be consistent with rocsolver



Ideally Tensile should be a separate package but that can be handled in a different PR.

teonnik committed:
rocBLAS: Add an option to disable tensile

- Change variant name to be consistent with rocsolver
- Change default variant value to be consistent with rocsolver

srekolam commented:
@cgmb  ,can you please review it

cgmb commented:
Does rocSOLVER build with `rocblas~tensile`? IIRC, there used to be missing symbols if you tried to do that.

teonnik commented:
@cgmb 

> Does rocSOLVER build with rocblas~tensile? IIRC, there used to be missing symbols if you tried to do that.

The package builds fine. However, I checked for missing symbols and indeed `gemm`s are missing. I thought that `rocblas` would have those even without `tensile`, I guess I was wrong. :/

haampie commented:
@cgmb do you think rocblas could have a fallback gemm that's quick to build?

cgmb commented:
That's been on my personal wishlist for a while... but I have a rather long wishlist. My suggestion would be to open an issue requesting the feature from rocBLAS.

cgmb commented on var/spack/repos/builtin/packages/rocblas/package.py:
```diff
@@ -29,27 +29,28 @@ class Rocblas(CMakePackage):
+                      'gfx1012', 'gfx1030')
 
-    variant('tensile_architecture', default='all', values=tensile_architecture, multi=True)
+    variant('amdgpu_target', default='gfx906:xnack-', values=amdgpu_targets, multi=True)
```
I really don't like defaulting to one random architecture. Yeah, I have gfx906 in my workstation, but now everyone on W6800, MI100, MI200, etc. needs to manually specify the architecture. And they probably will only find out they need to do that by building everything, running their program, having it fail, and debugging what's missing.

teonnik replied:
That is a good point. I chose that architecture to match the one set in `rocsolver` and to reduce the compile time.

srekolam replied:
@teonnik  Tensile is already in a separate package called 
rocm-tensile.

teonnik replied:
@srekolam Thanks! I missed that one! I will try to rewrite the PR in terms of that package.

cgmb replied:
I'm not sure that makes sense. Tensile exists in three forms:
1. It's a command-line executable for generating kernels, benchmarking them, and saving the best solutions in YAML files.
2. It's a build system component that reads YAML solution files, generates the specified kernels, and invokes the compiler to build them. It stores the results of that process in TensileLibrary.dat, and various code object files (`*.co`, `*.hsaco`).
3. It's a run-time library that will load the TensileLibrary.dat and code object files when you call it to do a GEMM, and choose the ideal kernel for your specific parameters.

I could be totally wrong, but I would think rocm-tensile only fulfills role (1). In any case, the YAML solution files are a part of rocBLAS, not Tensile (i.e., they are solutions to rocBLAS's problems). You won't save any time building Tensile separately, because (2) is what takes so long, and it needs rocBLAS's data. So, there's not much benefit to separating them.

I'm also not sure if (3) can be built ahead of time. @lringham, can you use the built version of rocm-tensile to build rocBLAS, or does it _require_ the Tensile source? For reference, [this is how Tensile is being built in the separate package](https://github.com/spack/spack/blob/66c70b82927fc7d3e37009461407fe282873a9a9/var/spack/repos/builtin/packages/rocm-tensile/package.py).

cgmb replied:
I forwarded this discussion to one of the senior members of the rocBLAS team, who responded, "if they build rocBLAS and Tensile differently than us, [...] it will be a maintenance problem to support them."

I would not recommend using a separate rocm-tensile.

teonnik replied:
@cgmb OK, I will not try to use `rocm-tensile` as a dependency in this package. I will also try to return the previous previous behavior which was to build all architectures by default. I would like to do that without an explicit `all` value but I am not sure if that is possible at the moment. I will have a go at it next week.

cgmb replied:
> I will also try to return the previous previous behavior which was to build all architectures by default. I would like to do that without an explicit all value but I am not sure if that is possible at the moment. I will have a go at it next week.

By default, all the ROCm math libraries will build for all the architectures that they support when `-DAMDGPU_TARGETS` is not passed. IMO, until Spack has proper support for detecting and setting GPU architectures, this should be the default.

IMO, don't worry about the 'all' thing in this PR. Just keep the existing behaviour. Then we open another PR that fixes the defaults and adds a consistent 'default' variant for all the ROCm math libraries.

cgmb replied:
I opened #28907 for the `AMDGPU_TARGETS` defaults and propagation change.

teonnik replied:
@cgmb  I see that you have incorporated most of the changes from this PR. Perhaps it is better to continue on your PR and close this one?

cgmb commented:
I haven't actually tested it, but I like this change aside from the default amdgpu_target.

cgmb commented:
There will be some minor merge conflicts to resolve if #27436 is merged first, but they should be pretty easy to deal with.

cgmb commented on var/spack/repos/builtin/packages/rocblas/package.py:
```diff
@@ -29,27 +29,28 @@ class Rocblas(CMakePackage):
 
-    variant('tensile_architecture', default='all', values=tensile_architecture, multi=True)
+    variant('amdgpu_target', default='gfx906:xnack-', values=amdgpu_targets, multi=True)
+    variant('tensile', default=False, description='Use Tensile as a backend')
```
I didn't notice this was defaulted to `False` until I went to use it, but it should be `True`. The library won't function as documented without Tensile. It's a feature that exists mostly for specific sorts of testing and debugging.

cgmb reviewed (changes requested):
The defaults on the variants are not appropriate. This should not be merged until they're changed.

By the way, there are build speed improvements coming for Tensile in ROCm 5.0/5.1, so you do have something to look forward to.

teonnik committed:
Merge branch 'develop' into disable_tensile

teonnik committed:
Set default for tensile to True to avoid surprises with missing symbols

teonnik committed:
Use auto_or_any_combination_of for AMD GPU architecture variant

teonnik commented:
I addressed the comments regarding the default variant and copied the changes for the variant from #28907 for rocblas. This now has the same changes for rocblas as in #28907 except that the default for tensile is set to True. If merging this PR before #28907 is preferable, then this can be merged, if not, then this PR can be closed and only #28907 can be merged (after the default value is changed there).

cgmb reviewed (approved):
This change still needs testing to verify it doesn't break anything, but the code looks good.

cgmb commented:
> There will be some minor merge conflicts to resolve if #27436 is merged first, but they should be pretty easy to deal with.

It has been merged, so the conflicts will need to be resolved.

cgmb commented:
This PR was closed because it was subsumed into #28907.

I talked with the rocBLAS team about this and they were receptive to fleshing out the `BUILD_WITH_TENSILE=OFF` variant into something more usable by third-parties. The optimized Tensile kernels are important for achieving peak performance with rocblas, but they do introduce build-time and run-time complexity that's not always needed.
